### PR TITLE
Setup api

### PR DESCRIPTION
**Pull request by GameSetAndMatch**
Setup app scripts to allow the use of a REST API to serve AddressParser Models

Docker-compose file added along with app definition

### CODE REVIEW

File: deepparse/app/app.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+address_parser_mapping: Dict[str, AddressParser] = {}
+
+configure_sentry()
+
+
+@asynccontextmanager
+async def lifespan(application: FastAPI):  # pylint: disable=unused-argument
+    # Load the models
+    logger.debug("downloading models")
```
Comment (davebulaval) [inline deepparse/app/app.py:27]:
```suggestion
    logger.debug("Downloading models")
```

File: pytest.ini [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [pytest]
-addopts = --cov ./deepparse --cov-report html --cov-report xml --cov-config=.coveragerc
+#addopts = --cov ./deepparse --cov-report html --cov-report xml --cov-config=.coveragerc
```
Comment (davebulaval) [inline pytest.ini:2]: Reason for this?

File: requirements.txt [hunk 3]
```diff
@@ -10,3 +10,7 @@ fasttext-wheel
 pandas
 urllib3
 cloudpathlib[s3, gs, azure]
+fastapi[all]
+uvicorn
```
Comment (davebulaval) [inline requirements.txt:14]: I think the four new requirements should be more in the extra requirements and be in an api_requirements.txt for developers, and rather be downloaded on the side.

File: setup.py [hunk 4]
```diff
@@ -80,6 +80,10 @@ def main():
             "pandas",
             "urllib3",
             "cloudpathlib[s3, gs, azure]",
+            "fastapi[all]",
+            "uvicorn",
+            "sentry-sdk[fastapi]",
```
Comment (davebulaval) [inline setup.py:85]: See the previous comment.

File: pytest.ini (hunk 2 above)
Comment (GameSetAndMatch) [reply pytest.ini:2]: Forgot to remove, you cannot debug tests without this line commented

File: requirements.txt (hunk 3 above)
Comment (GameSetAndMatch) [reply requirements.txt:14]: I see the point, should the app be in another package from deepparse then ? And it has deepparse package as a dependency ?

Comment (GameSetAndMatch) [reply requirements.txt:14]: Otherwise people can and will try to import the app without the proper packages installed

File: setup.py (hunk 4 above)
Comment (GameSetAndMatch) [reply setup.py:85]: See my previous answer

File: requirements.txt (hunk 3 above)
Comment (davebulaval) [reply requirements.txt:14]: No, we just need to add a try/catch in the app's import and raise a message if the imports are not there.

Comment (davebulaval) [reply requirements.txt:14]: Look at colorama import.

Comment (GameSetAndMatch) [reply requirements.txt]: After a conversation with @davebulaval, we concluded that the best way to import the app dependencies were to pass them as extra requirements, this way the app dependencies can be installed using `pip install deepparse[app]`

File: deepparse/app/app.py [hunk 5]
```diff
@@ -15,9 +15,7 @@
 
 
 except ModuleNotFoundError as e:
-    raise ModuleNotFoundError(
-        "Ensure you installed the packages for the app_requirements.txt file found in the root of the project"
-    ) from e
+    raise ModuleNotFoundError("Ensure you installed the extra packages using: 'pip install deepparse[app]''") from e
```
Comment (davebulaval) [inline deepparse/app/app.py:18]:
```suggestion
    raise ModuleNotFoundError("Ensure you installed the extra packages using: 'pip install deepparse[app]'") from e
```

**Review by davebulaval — APPROVED** [review]

**Comment by davebulaval** [comment]
It misses a doc in the README.md and index.rst. I have placed a placeholder where I see it. It also misses a .env_example to raise the app.

**Comment by GameSetAndMatch** [comment]
Added documentation

**Review by davebulaval — APPROVED** [review]

**Comment by davebulaval** [comment]
Is it possible de deactivate sentry in the docker-compose?

**Comment by GameSetAndMatch** [comment]
Well it is possible to just not use it.

Lets say someone set the SENTRY_DSN env variable set to nothing (`SENTRY_DSN=`) it will just bypass it and not use Sentry logging. Would that be a good workaround for someone not willing to use it?

File: README.md [hunk 6]
```diff
@@ -246,6 +244,18 @@ Once you have Docker Engine and Docker Compose installed, you can run the follow
...
+DSN. There is an example of the .env file in the root of the project named `.env_example`. you can just copy it using the following command: 
+
+```shell
+cp .env_example .env
+```
+
+If you do not have a Sentry account, you can create one [here](https://sentry.io/signup/).
+
+If you do not want to use Sentry, you can just remove the `SENTRY_DSN` environment variable from the `.env` file or set it to an empty string, also, not creating
+the `.env` file will also work. The application will run without any problem if Sentry is not set.
```
Comment (davebulaval) [inline README.md:258]:
```suggestion
#### Sentry:
Also, you can monitor your application usage with [Sentry](https://sentry.io) by setting the environment variable `SENTRY_DSN` to your Sentry's project
DSN. There is an example of the .env file in the project's root named `.env_example`. You can copy it using the following command: 

```shell
cp .env_example .env
```

If you do not have a Sentry account, you can create one [here](https://sentry.io/signup/).

If you do not want to use Sentry, you can remove the `SENTRY_DSN` environment variable from the `.env` file or set it to an empty string; also, not creating
the `.env` file will work. The application will run without any problem if Sentry is not set.
```

File: deepparse/app/app.py [hunk 7]
```diff
@@ -33,7 +34,7 @@ async def lifespan(application: FastAPI):  # pylint: disable=unused-argument
     logger.debug("Downloading models")
     download_models()
     for model in MODEL_MAPPING_CHOICES:
-        if model not in ["fasttext", "fasttext-attention", "fasttext-light"]:
+        if model not in ["fasttext", "fasttext-attention", "fasttext-light"]:  # Skip fasttext models
```
Comment (davebulaval) [inline deepparse/app/app.py:37]: Reason to skip?

**Review by davebulaval — APPROVED** [review]

File: deepparse/app/app.py (hunk 7 above)
Comment (GameSetAndMatch) [reply deepparse/app/app.py:37]: Thought I answered this one, sorry. No reason to skip when the PR is ready to merge.

**Pull request merged**